comms/storage/logstream/logstream.go

```diff
@@ -0,0 +1,161 @@
+	Host   string    `json:"host"`
+	LastOK time.Time `json:"lastOk"`
+	Shards []string  `json:"shards"`
+}
+
+type UpdateHealthyNodeSetLog struct {
+	HealthyNodes []string `json:"healthyNodes"`
+	UpdatedBy    string   `json:"updatedBy"`
+}
+
+type RebalanceLog struct {
+	PrevShards []string `json:"prevShards"`
+	NewShards  []string `json:"newShards"`
+}
+
+func New(namespace string, pubKey string, jsc nats.JetStreamContext) (*LogStream, error) {
+	err := retry.Do(func() error {
+		_, err := jsc.AddStream(&nats.StreamConfig{
+			Name:        StreamName,
+			Description: "Core actions logged by each node",
+			Subjects:    []string{"storage.log.*.statusUpdate", "storage.log.updateHealthyNodeSet", "storage.log.*.rebalanceStart", "storage.log.*.rebalanceEnd"},
+			Replicas:    3,
+		})
+		return err
+	}, retry.Attempts(5))
+	if err != nil {
+		return nil, err
+	}
+
+	return &LogStream{
+		namespace: namespace,
+		pubKey:    pubKey,
+		jsc:       jsc,
+	}, nil
+}
+
+func (ls LogStream) LogStatusUpdate(status []byte) {
+	err := retry.Do(func() error {
+		_, err := ls.jsc.Publish(fmt.Sprintf("storage.log.%s.statusUpdate", ls.pubKey), status)
+		return err
+	})
+	if err != nil {
+		slog.With("pubKey", ls.pubKey).Warn("failed to record status update: %v", err)
+	}
+}
+
+func (ls LogStream) LogUpdateHealthyNodeSet(healthyNodes []string) {
+	msg, err := json.Marshal(UpdateHealthyNodeSetLog{
+		HealthyNodes: healthyNodes,
+		UpdatedBy:    ls.pubKey,
+	})
+	if err != nil {
+		slog.With("pubKey", ls.pubKey).Warn("failed to marshal data to record updateHealthyNodeSet: %v", err)
+	}
+	_, err = ls.jsc.Publish("storage.log.updateHealthyNodeSet", msg)
+	if err != nil {
+		slog.With("pubKey", ls.pubKey).Warn("failed to record updateHealthyNodeSet: %v", err)
+	}
+}
+
+func (ls LogStream) LogRebalanceStart(prevShards []string, newShards []string) {
+	msg, err := json.Marshal(RebalanceLog{
+		PrevShards: prevShards,
+		NewShards:  newShards,
+	})
+	if err != nil {
+		slog.With("pubKey", ls.pubKey).Warn("failed to marshal data to record rebalanceStart: %v", err)
+	}
+	_, err = ls.jsc.Publish(fmt.Sprintf("storage.log.%s.rebalanceStart", ls.pubKey), msg)
+	if err != nil {
+		slog.With("pubKey", ls.pubKey).Warn("failed to record rebalanceStart: %v", err)
+	}
+}
+
+func (ls LogStream) LogRebalanceEnd(prevShards []string, newShards []string) {
+	msg, err := json.Marshal(RebalanceLog{
+		PrevShards: prevShards,
+		NewShards:  newShards,
+	})
+	if err != nil {
+		slog.With("pubKey", ls.pubKey).Warn("failed to marshal data to record rebalanceEnd: %v", err)
+	}
+	_, err = ls.jsc.Publish(fmt.Sprintf("storage.log.%s.rebalanceEnd", ls.pubKey), msg)
+	if err != nil {
+		slog.With("pubKey", ls.pubKey).Warn("failed to record rebalanceEnd: %v", err)
+	}
+}
+
+// GetNStatusUpdatesSince returns the first N status updates for a node that occurred on or after numHoursAgo hours ago.
+func (ls LogStream) GetNStatusUpdatesSince(n int, numHoursAgo float64, pubKey string) ([]NodeStatus, error) {
```

Comment:
since nats streams are append-only logs rather than time series, the way I do a time-based query on the stream is to get N logs starting at X hours ago:
* make an ordered consumer starting at X hours ago
* read events until we've seen N log events or it until times out (i.e., no more new log events - end of stream)
* return the N events

then the frontend has a datetime picker and requests N log events starting at X hours ago, where X=(nowTime-desiredTime).hours()